src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java

```diff
@@ -0,0 +1,251 @@
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Throwables;
+import org.apache.http.HttpStatus;
+import org.apache.http.client.utils.URIBuilder;
+import org.embulk.config.ConfigException;
+import org.embulk.config.TaskReport;
+import org.embulk.input.zendesk.RecordImporter;
+import org.embulk.input.zendesk.ZendeskInputPlugin;
+import org.embulk.input.zendesk.clients.ZendeskRestClient;
+import org.embulk.input.zendesk.models.ZendeskException;
+import org.embulk.input.zendesk.utils.ZendeskConstants;
+import org.embulk.input.zendesk.utils.ZendeskDateUtils;
+import org.embulk.input.zendesk.utils.ZendeskUtils;
+import org.embulk.spi.Exec;
+import org.slf4j.Logger;
+
+import java.time.Instant;
+
+import java.util.Iterator;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+public abstract class ZendeskNormalServices implements ZendeskService
+{
+    private static final Logger logger = Exec.getLogger(ZendeskNormalServices.class);
+
+    protected ZendeskInputPlugin.PluginTask task;
+
+    private ZendeskRestClient zendeskRestClient;
+
+    protected ZendeskNormalServices(final ZendeskInputPlugin.PluginTask task)
+    {
+        this.task = task;
+    }
+
+    protected abstract String buildURI(int page, long startTime);
+
+    @VisibleForTesting
+    protected ZendeskRestClient getZendeskRestClient()
+    {
+        if (zendeskRestClient == null) {
+            zendeskRestClient = new ZendeskRestClient();
+        }
+        return zendeskRestClient;
+    }
+
+    public TaskReport execute(final int taskIndex, final RecordImporter recordImporter)
+    {
+        TaskReport taskReport = Exec.newTaskReport();
+
+        if (isSupportIncremental()) {
+            importDataForIncremental(task, recordImporter, taskReport);
+        }
+        else {
+            importDataForNonIncremental(task, taskIndex, recordImporter);
+        }
+
+        return taskReport;
+    }
+
+    public JsonNode getData(String path, final int page, final boolean isPreview, final long startTime)
+    {
+        if (path.isEmpty()) {
+            path = buildURI(page, startTime);
+        }
+        try {
+            final String response = getZendeskRestClient().doGet(path, task, isPreview);
+            return ZendeskUtils.parseJsonObject(response);
+        }
+        catch (final Exception e) {
+            throw Throwables.propagate(e);
+        }
+    }
+
+    protected URIBuilder getURIBuilderFromHost()
+    {
+        return ZendeskUtils.getURIBuilder(task.getLoginUrl());
+    }
+
+    private void importDataForIncremental(final ZendeskInputPlugin.PluginTask task, final RecordImporter recordImporter, final TaskReport taskReport)
+    {
+        long startTime = 0;
+        if (!Exec.isPreview() && isSupportIncremental()) {
```

Comment:
I have few points here:
1. In the original code snippet, `Exec.isPreview` is not checked: https://github.com/treasure-data/embulk-input-zendesk/blob/d669a2c063ecbb17e141254d8be68b7fca00f399/src/main/java/org/embulk/input/zendesk/ZendeskInputPlugin.java#L249-L251, I'm not sure why it brought here, and why `startTime` should not be updated if it is not in preview phase?
2. I think `isSupportIncremental` is redundant, I'd assume the method `importDataForIncremental` is able to call in incremental mode only.

Reply:
fixed